gridcomps/History/MAPL_StationSamplerMod.F90

```diff
@@ -0,0 +1,461 @@
+       call v%add_attribute('valid_range',(/-MAPL_UNDEF,MAPL_UNDEF/))
+       call this%fmd%add_variable(trim(var_name),v,_RC)
+    enddo
+    deallocate (fieldNameList)
+    
+    ! locstream route handle
+    call ESMF_FieldBundleGet(bundle,grid=grid,_RC)
+    this%regridder = LocStreamRegridder(grid,this%esmf_ls,_RC)
+    _RETURN(_SUCCESS)
+  end subroutine add_metadata_route_handle
+
+  
+  subroutine append_file(this,current_time,rc)
+     class(StationSampler), intent(inout) :: this
+     type(ESMF_Time), intent(in) :: current_time
+     integer, optional, intent(out) :: rc
+     !
+     integer :: status
+     integer :: fieldCount
+     integer :: ub(1), lb(1)
+     type(ESMF_Field) :: src_field,dst_field
+     real(kind=REAL32), allocatable :: p_new_lev(:,:,:)
+     real(kind=REAL32), pointer :: p_src_3d(:,:,:),p_src_2d(:,:)
+     real(kind=REAL32), pointer :: p_dst_3d(:,:),p_dst_2d(:)
+     real(kind=REAL32), allocatable :: arr(:,:)
+     character(len=ESMF_MAXSTR), allocatable ::  fieldNameList(:)
+     character(len=ESMF_MAXSTR) :: xname
+     real(kind=ESMF_KIND_R8), allocatable :: rtimes(:)
+     integer :: i, id, iobs, ix, rank
+     integer :: nx, nz
+     
+     this%obs_written=this%obs_written+1
+     
+     !__ s1. put_var: time variable
+     !!call ESMF_TimePrint(current_time,options='string')
+     rtimes = this%compute_time_for_current(current_time,_RC) ! rtimes: seconds since opening file
+     if (this%vdata%regrid_type==VERTICAL_METHOD_ETA2LEV) then
+        call this%vdata%setup_eta_to_pressure(_RC)
+     end if
+     if (mapl_am_i_root()) then
+        call this%formatter%put_var('time',rtimes(1:1),&
+             start=[this%obs_written],count=[1],_RC)
+     end if
+
+     !__ s2. put_var: ungridded_dim from src to dst [interpolation]
+     call ESMF_FieldBundleGet(this%bundle, fieldCount=fieldCount, _RC)
+     allocate (fieldNameList(fieldCount))
+     call ESMF_FieldBundleGet(this%bundle, fieldNameList=fieldNameList, _RC)
+     do i=1, fieldCount
+        xname=trim(fieldNameList(i))
+        call ESMF_FieldBundleGet(this%bundle,xname,field=src_field,_RC)
+        call ESMF_FieldGet(src_field,rank=rank,_RC)
+        if (rank==2) then
+           call ESMF_FieldGet(src_field,farrayptr=p_src_2d,_RC)
+           dst_field = ESMF_FieldCreate(this%esmf_ls,name=xname, &
+                typekind=ESMF_TYPEKIND_R4,_RC)
+           call ESMF_FieldGet(dst_field,farrayptr=p_dst_2d,_RC)
+           call this%regridder%regrid(p_src_2d,p_dst_2d,_RC)
+           if (mapl_am_i_root()) then
+              call this%formatter%put_var(xname,p_dst_2d,&
+                   start=[1,this%obs_written],count=[this%nstation,1],_RC)                         
+           end if
+           call ESMF_FieldDestroy(dst_field,nogarbage=.true.)
+        else if (rank==3) then
+           call ESMF_FieldGet(src_field,farrayptr=p_src_3d,_RC)
+           call ESMF_FieldGet(src_field,ungriddedLBound=lb,ungriddedUBound=ub,_RC)
```

Comment:
Consider moving this entire block of code into a helper procedure.   Will shorten the main procedure.

Reply:
This section to write to nc file contains 42 lines of code.   The whole subroutine append_file contains ~80 lines of code.  I still prefer to keep the flow without shifting into a new subroutine. (Maybe that is because I am getting used to the dense block.)